File: notification_service/src/notification/domain/entity.py
from dataclasses import dataclass
from typing import Any, Self
from uuid import UUID, uuid4

from fern_labour_core.aggregate_root import AggregateRoot

from src.notification.domain.enums import (
    NotificationChannel,
    NotificationStatus,
)
from src.notification.domain.events import NotificationCreated, NotificationStatusUpdated
from src.notification.domain.value_objects.notification_id import NotificationId


@dataclass(eq=False, kw_only=True)
class Notification(AggregateRoot[NotificationId]):
    status: NotificationStatus
    channel: NotificationChannel
    destination: str
    template: str
    data: dict[str, Any]
    metadata: dict[str, Any] | None = None
    external_id: str | None = None

    @classmethod
    def create(
        cls,
        *,
        channel: NotificationChannel,
        destination: str,
        template: str,
        data: dict[str, Any],
        status: NotificationStatus | None = None,
        metadata: dict[str, Any] | None = None,
        external_id: str | None = None,
        notification_id: UUID | None = None,
    ) -> Self:
        notification_id = notification_id or uuid4()
        notification = cls(
            id_=NotificationId(notification_id),
            status=status or NotificationStatus.CREATED,
            channel=channel,
            destination=destination,
            template=template,
            data=data,
            metadata=metadata,
            external_id=external_id,
        )
        notification.add_domain_event(
            NotificationCreated.create(
                aggregate_id=str(notification_id),
                aggregate_type="notification",
                data={"notification_id": str(notification_id)},
            )
        )
        return notification

    def update_status(self, status: NotificationStatus) -> None:
        self.add_domain_event(
            NotificationStatusUpdated.create(
                aggregate_id=str(self.id_.value),
                aggregate_type="notification",
                data={
                    "notification_id": str(self.id_.value),
                    "channel": self.channel.value,
                    "external_id": self.external_id,
                    "from_status": self.status.value,
                    "to_status": status.value,
                },
            )
        )
        self.status = status

File: rust_notification_service/src/domain/entity.rs
use std::collections::HashMap;

use serde::{Deserialize, Serialize};
use uuid::Uuid;

use super::enums::{NotificationChannel, NotificationStatus, NotificationTemplate};

#[derive(Debug, Deserialize, Serialize, Clone)]
pub struct Notification {
    pub id: Uuid,
    pub status: NotificationStatus,
    pub channel: NotificationChannel,
    pub destination: String,
    pub template: NotificationTemplate,
    pub data: HashMap<String, String>,
    pub metadata: Option<HashMap<String, String>>,
    pub external_id: Option<String>,
}

impl Notification {
    pub fn create(
        channel: NotificationChannel,
        destination: String,
        template: NotificationTemplate,
        data: HashMap<String, String>,
        status: Option<NotificationStatus>,
        metadata: Option<HashMap<String, String>>,
    ) -> Self {
        return Self {
            id: Uuid::now_v7(),
            status: status.unwrap_or(NotificationStatus::CREATED),
            channel,
            destination,
            template,
            data,
            metadata,
            external_id: None,
        };
    }

    pub fn set_external_id(&mut self, external_id: String) {
        self.external_id = Some(external_id);
    }

    pub fn set_status(&mut self, status: NotificationStatus) {
        self.status = status;
    }
}
